pkg/webhook/elasticquota/quota_topology_check.go

```diff
@@ -0,0 +1,259 @@
+	"github.com/koordinator-sh/koordinator/pkg/scheduler/plugins/elasticquota/core"
+)
+
+func (qt *quotaTopology) validateQuotaSelfItem(quota *v1alpha1.ElasticQuota) error {
+	if invalid, err := extension.IsForbiddenModify(quota); invalid {
+		return err
+	}
+
+	// min and max's each dimension should not have negative value
+	if resourceNames := quotav1.IsNegative(quota.Spec.Max); len(resourceNames) > 0 {
+		return fmt.Errorf("%v quota.Spec.Max's value < 0, in dimensions :%v", quota.Name, resourceNames)
+	}
+
+	if resourceNames := quotav1.IsNegative(quota.Spec.Min); len(resourceNames) > 0 {
+		return fmt.Errorf("%v quota.Spec.Min's value < 0, in dimensions :%v", quota.Name, resourceNames)
+	}
+
+	var sharedRatio v1.ResourceList
+	if quota.Annotations[extension.AnnotationSharedWeight] != "" {
+		if err := json.Unmarshal([]byte(quota.Annotations[extension.AnnotationSharedWeight]), &sharedRatio); err != nil {
+			return err
+		}
+
+		if resourceNames := quotav1.IsNegative(sharedRatio); len(resourceNames) > 0 {
+			return fmt.Errorf("%v quota.Annotation[%v]'s value < 0, in dimension :%v", quota.Name, extension.AnnotationSharedWeight, resourceNames)
+		}
+	}
+
+	// minQuota <= maxQuota
+	for key, val := range quota.Spec.Min {
+		if maxVal, exist := quota.Spec.Max[key]; !exist || maxVal.Cmp(val) == -1 {
+			return fmt.Errorf("%v min :%v > max,%v", quota.Name, quota.Spec.Min, quota.Spec.Max)
+		}
+	}
+	return nil
+}
+
+// validateQuotaTopology checks the quotaInfo's topology with its parent and its children (when update calls the function).
+func (qt *quotaTopology) validateQuotaTopology(oldQuotaInfo, quotaInfo *core.QuotaInfo) error {
+	// interchange between parentQuotaGroup and childQuotaGroup is prohibited now.
+	if oldQuotaInfo != nil && oldQuotaInfo.IsParent != quotaInfo.IsParent {
+		return fmt.Errorf("IsParent is forbidden modify now, quotaName:%v", oldQuotaInfo.Name)
+	}
+
+	// if the quotaInfo's parent is root and its IsParent is false, the following checks will be true, just return nil.
+	if quotaInfo.ParentName == extension.RootQuotaName && !quotaInfo.IsParent {
+		return nil
+	}
+
+	if err := qt.checkParentQuotaInfoExist(quotaInfo.Name, quotaInfo.ParentName); err != nil {
+		return err
+	}
+
+	if err := qt.checkSubAndParentGroupMaxQuotaKeySame(quotaInfo); err != nil {
+		return err
+	}
+
+	if err := qt.checkMinQuotaSum(quotaInfo); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (qt *quotaTopology) checkParentQuotaInfoExist(quotaName, parentName string) error {
+	if parentName != extension.RootQuotaName {
+		parentInfo, find := qt.quotaInfoMap[parentName]
+		if !find {
+			return fmt.Errorf("%v has parentName %v but not find parentInfo in quotaInfoMap", quotaName, parentName)
+		}
+		if _, exist := qt.quotaHierarchyInfo[parentName]; !exist {
+			return fmt.Errorf("%v has parentName %v but not find parentInfo in quotaHierarchyInfo", quotaName, parentName)
+		}
+		if !parentInfo.IsParent {
```

Comment:
if !parentInfo.IsParent{} should not appear in  "checkParentQuotaInfoExist", give a new func may be better

Reply:
fixed